frontend/packages/operator-lifecycle-manager/src/components/create-operand.tsx

```diff
@@ -675,465 +919,376 @@ export const CreateOperandForm: React.FC<CreateOperandFormProps> = ({
+    // Flat array of fields in this arrayFieldGroup at index > removed.
+    const fieldsToLeftShift = _.flatten(
+      _.filter(fieldLists, (_unused, index) => index > indexToRemove),
+    );
 
-  const addGroup = (group) => {
-    const groupList = _.cloneDeep(arrayFieldGroups);
+    // List of paths to be removed from fields list. Fields and values in fieldsToLeftShift
+    // will be replaced with left-shifted equivalents
+    const pathsToRemove = _.map([...fieldLists[indexToRemove], ...fieldsToLeftShift], 'path');
+
+    // Left-shift fields and formValues that were subsequent to the removed arrayFieldGroup
+    const [leftShiftedFields, leftShiftedValues] = _.reduce(
+      fieldsToLeftShift,
+      ([fieldAccumulator, formValueAccumulator], field) => {
+        const path = modifyArrayFieldPathIndex(field.path, (index) => index - 1);
+        return [
+          [...fieldAccumulator, { ...field, path }],
+          { ...formValueAccumulator, [path]: _.get(formValues, field.path) },
+        ];
+      },
+      [[], {}],
+    );
 
-    if (!_.isNil(group) && arrayFieldGroups.size > 0) {
-      groupList.add(group);
-      setArrayFieldGroupList(new Set([...groupList].sort()));
-    }
-  };
+    // Remove arrayFieldGroup from fields and replace subsequent arrayFieldGroup fields with
+    // left-shifted ones.
+    setFields((currentFields) => {
+      return [
+        // Filter out all removed fields and fields that need to be left-shifted
+        ..._.filter(currentFields, (field) => !_.includes(pathsToRemove, field.path)),
+        // Add new left-shifted fields
+        ...leftShiftedFields,
+      ];
+    });
 
-  const removeGroup = (groupName) => {
-    const groupList = [...arrayFieldGroups];
-    const newGroupList = groupList.filter((g) => g.includes(groupName));
+    // Remove arrayFieldGroup from formValues and replace subsequent arrayFieldGroup values with
+    // the left-shifted ones.
+    setFormValues((currentValues) => {
+      // Filter out all removed values and values that need to be left-shifted
+      return {
+        ..._.omit(currentValues, pathsToRemove),
 
-    if (!_.isNil(groupName) && !_.isEmpty(groupList) && !_.isEmpty(newGroupList)) {
-      const newList = _.without(groupList, _.last(newGroupList));
-      setArrayFieldGroupList(new Set(newList.sort()));
-    }
+        // Add new left-shifted values
+        ...leftShiftedValues,
+      };
+    });
   };
 
-  const addGroupAndFields = (fieldList, groupName) => {
-    const groupIndex = _.filter([...arrayFieldGroups], (g) =>
-      g.split(':').includes(groupName),
-    ).length.toString();
-    const newGroup = SpecCapability.arrayFieldGroup
-      .concat(groupName)
-      .concat(':')
-      .concat(groupIndex);
-
-    if (!_.isNil(groupName) && !_.isEmpty(fieldList)) {
-      addGroup(newGroup);
-      addFieldsToGroup(fieldList, newGroup);
-    }
-  };
+  const renderArrayFieldGroups = () =>
```

Comment:
Just pulled this out of the jsx in the return statement to make it easier to read.

Reply:
All of these render helpers should probably be pulled out into separate components in the future.